### PR TITLE
[Product Editor] Variation quick actions succeed when sku is inherited from parent

### LINKED ISSUE #43800: [Product Editor] Variation quick actions fail when the variation does not have an explicit (unique) SKU set
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

In the new product editor, when you create a variable product and generate variations, those variations by default inherit the SKU of the variable product. However, the `quick update` actions for variations does not successfully update the variation because:

- When making a `POST` request to the REST API endpoint `/wc/v3/products/{PRODUCT_ID}/variations/{VARIATION_ID}`, we send the SKU, which results in a `400` error, since the SKU is not unique

Note: when editing a variation by clicking `Edit` and then make a change in the editor, we don't send the SKU, thus allowing the variation to be updated.

### Expected behavior

The `quick update` actions for variations are able to successfully update the variation.

### Actual behavior

The `quick update` actions for variations fail with a `Failed to save variation` toast message.

In the network tab of the browser dev tools, you will see a `400` response.

### Steps to reproduce

With a WooCommerce dev env with the new product editor enabled (**WooCommerce** > **Settings** > **Advanced** > **Features** > **Experimental features**)...

1. Go to **Products** > **Add New**
2. In **Variations** tab, generate some variations
3. Save draft
4. In list of variations, click on the **...** for a variation and select the **Pricing** > **Set list price** 
5. Enter a value and click **OK**
6. Verify the variation fails to be updated

Screenshot 2024-01-18 at 14 38 00

Screenshot 2024-01-18 at 14 36 52

### WordPress Environment

n/a

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by mattsherman**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR addresses the following:

- Variation quick actions (from either the "quick update" menu, or the "..." menu) succeed when the variation inherits its sku from the parent variable product (closes #43800)
- TypeScript updates to require the `id` property in the codebase that handles variation quick actions (since without an `id`, you can't identify what variation to update)
- Always use arrays to pass selection and changes in variation quick actions, to simplify handling (no more `if Array.isArray( selection)` checks everywhere!)

#### Automated testing

1. Make sure unit tests pass...

```bash
pnpm --filter=@woocommerce/product-editor test:js
```

#### Manual testing

With a WooCommerce dev env with the new product editor enabled (**WooCommerce** > **Settings** > **Advanced** > **Features** > **Experimental features**)...

1. Go to **Products** > **Add New**
2. Add some variations on the **Variations** tab
3. Verify quick actions (both "quick update" and "...") work as expected, properly updating the variations (these should work whether the variation has an explicit sku set or whether it inherits it from the variable parent)

Screenshot 2024-01-23 at 20 55 38
Screenshot 2024-01-23 at 20 55 27

4. Install the following test plugin, which adds two menu items under the "Pricing" group, and verify that the Slot Fills recently added in #43441 function:

- **Add 100 to regular price (single selection only)**
- **50% off sale (multiple selection allowed)**

[test-44017.zip](https://github.com/woocommerce/woocommerce/files/14054589/test-44017.zip)

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/components/variations-table/variation-actions-menus/test/variation-actions.test.tsx [hunk 1]
```diff
@@ -23,7 +23,7 @@ const mockVariation = {
 	downloads: [],
 	name: '',
 	parent_id: 1,
-} as ProductVariation;
+} as unknown as ProductVariation;
```
Comment (louwie17) [inline packages/js/product-editor/src/components/variations-table/variation-actions-menus/test/variation-actions.test.tsx:26]: I guess TS started complaining about this?

File: packages/js/product-editor/src/components/variations-table/variations-table.tsx [hunk 2]
```diff
@@ -235,10 +237,10 @@ export const VariationsTable = forwardRef<
 			} );
 	}
 
-	function handleDeleteAll( values: Partial< ProductVariation >[] ) {
+	function handleDeleteAll( values: PartialProductVariation[] ) {
 		const now = Date.now();
 
-		onBatchDelete( values.map( ( variation ) => variation.id ) )
+		onBatchDelete( values )
```
Comment (louwie17) [inline packages/js/product-editor/src/components/variations-table/variations-table.tsx:243]: This change breaks the delete. We will have to update the `onBatchDelete` in the `useVariations` hook.
This is causing us to pass the entire variation object as part the `delete` batch, although the `delete` option only takes an array of ids: https://woocommerce.github.io/woocommerce-rest-api-docs/#batch-update-products

**Review by louwie17 — CHANGES_REQUESTED** [review]
Great work on this @mattsherman, this overall tested well, I just noticed one bug around batch deleting variations ( should be a simple fix ).
Otherwise I like the move to using arrays for everything 👍

File: packages/js/product-editor/src/components/variations-table/variation-actions-menus/test/variation-actions.test.tsx (hunk 1 above)
Comment (mattsherman) [reply packages/js/product-editor/src/components/variations-table/variation-actions-menus/test/variation-actions.test.tsx:26]: Yes

File: packages/js/product-editor/src/components/variations-table/variations-table.tsx (hunk 2 above)
Comment (mattsherman) [reply packages/js/product-editor/src/components/variations-table/variations-table.tsx:243]: Oops! Thanks for catching this. I have updated `onBatchDelete`. It is odd though... it looks to me like the invalidation after the delete in `onBatchDelete` should have failed previously, as it was trying to use the values passed in as variations. I don't have time to check this on `trunk` right now, but will when I get back from my appointment just so I better understand how the code used to work (or not). But, in any case, things should be fixed up now!

**Review by louwie17 — APPROVED** [review]
Thanks for the update @mattsherman, bulk deleting is testing well now. You got my approval :shipit:

**Pull request merged**